firebase-common/src/main/java/com/google/firebase/annotations/NonNullGeneric.java

```diff
@@ -0,0 +1,32 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.annotations;
+
+import android.os.Build;
+
+import java.lang.annotation.ElementType;
+
+/**
+ * Nullability meta annotation for generic parameters.
+ *
+ * <p>Required to annotate generics as using annotations at `TYPE_USE` is not allowed in API level <
```

Comment:
This is a shame. Is there no way to do this, e.g. with something like retrolambda?

Reply:
Unfortunately I don't think there is another way. The issue is  that `Foo<@Bar Baz>` is not representable in bytecode prior to java version 8, so there is nothing to desugar it to. Hence this workaround where we "pretend" to have annotated at TYPE_USE.

One downside of this approach is that mixed nullability is not expressible e.g.: `Map<@NonNull String, @Nullable Integer>`. But it is what it is.